core/string/src/display.rs

```diff
@@ -0,0 +1,45 @@
+//! Display implementations for [`crate::JsString`].
+use crate::{CodePoint, JsStr, JsStrVariant};
+use std::fmt;
+use std::fmt::Write;
+
+/// Display implementation for [`crate::JsString`] that escapes unicode characters.
+#[derive(Debug)]
+pub struct JsStrDisplayEscaped<'a> {
+    inner: JsStr<'a>,
+}
+
+impl fmt::Display for JsStrDisplayEscaped<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self.inner.variant() {
+            // SAFETY: `JsStrVariant::Latin1` is always valid utf8, so no need to check.
+            JsStrVariant::Latin1(v) => v
+                .iter()
+                .copied()
+                .map(char::from)
+                .try_for_each(|c| f.write_char(c)),
+            JsStrVariant::Utf16(_) => self.inner.code_points().try_for_each(|r| match r {
+                CodePoint::Unicode(c) => f.write_char(c),
+                CodePoint::UnpairedSurrogate(c) => {
+                    f.write_str("\\u")?;
+                    write!(f, "{c:04X}")
```

Comment:
```suggestion
                    write!(f, "\\u{c:04X}")
```

Is this an optimization? I think we can just inline the two calls.